Si702x.class.nut

```diff
@@ -1,73 +1,116 @@
+        } else if (hardware.millis() - startTime < TIMEOUT_MS) {
+            imp.wakeup(0, function() {
+                _pollForResult(startTime, callback);
+            }.bindenv(this));
+        } else {
+            // Timeout
+            callback(null);
         }
-        local humidity = RH_MULT*((reading[0] << 8) + reading[1]) + RH_ADD;
-        return humidity;
     }
 
-    // read the temperature
-    // Input: (none)
-    // Return: temperature in celsius (float)
-    function readTemp() {
-        _i2c.write(_addr, READ_TEMP);
-        local reading = _i2c.read(_addr, "", 2);
-        local start = hardware.millis();
-        while (reading == null && (hardware.millis() - start < TIMEOUT)) {
-            reading = _i2c.read(_addr, "", 2);
+    // Starts a relative humidity measurement
+    function _readRH(callback=null) {
+        _i2c.write(_addr, MEASURE_RH);
+        local startTime = hardware.millis();
+        if (callback == null) {
+            local result = _i2c.read(_addr, "", 2);
+            while (result == null && hardware.millis() - startTime < TIMEOUT_MS) {
+                result = _i2c.read(_addr, "", 2);
+            }
+            return result;
+        } else {
+            _pollForResult(startTime, callback);
         }
-        if (hardware.millis() - start >= TIMEOUT) {
-            throw "Timed out waiting for temperature response from SI7021";
+    }
+
+    // Reads and returns the temperature value from the previous humidity measurement
+    function _readTempFromPrev() {
+        local rawTemp = _i2c.read(_addr, READ_PREV_TEMP, 2);
+        if (rawTemp) {
+            return TEMP_MULT*((rawTemp[0] << 8) + rawTemp[1]) + TEMP_ADD;
+        } else {
+            server.log("Si702x i2c read error: " + _i2c.readerror());
+            return null;
         }
-        local temperature = TEMP_MULT*((reading[0] << 8) + reading[1]) + TEMP_ADD;
-        return temperature;
     }
 
-    // read the temperature from previous rh measurement
-    // this method does not have to recalculate temperature so it is faster
-    // Input: (none)
-    // Return: temperature in celsius (float)
-    function readPrevTemp() {
-        _i2c.write(_addr, PREV_TEMP);
-        local reading = _i2c.read(_addr, "", 2);
-        if (!reading) { throw "I2C Error reading previous temperature from SI7021"; }
-        local temperature = TEMP_MULT*((reading[0] << 8) + reading[1]) + TEMP_ADD;
-        return temperature;
+    // Initiates a relative humidity measurement,
+    // then passes the humidity and temperature readings as a table to the user-supplied callback, if it exists
+    // or returns them to the caller, if it doesn't
+    function read(callback=null) {
+        if (callback == null) {
+            local rawHumidity = _readRH();
+            local temp = _readTempFromPrev();
+            if (rawHumidity == null || temp == null) {
+                server.log("Si702x reading timed out");
+                return null;
```

Comment:
I'm a fan of consistency - so I would say a table identical to what would be passed into the callback.